### PR TITLE
[UNR-5037] Modular client server rpcs

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Read in commit order.

This PR adds the Sender/Receiver for RingBuffer RPC with ACKs, and wire them up to the NetDriver.
It also add callbacks for RPC queues to report some error state, which might differ in what they can report, and what we want to do depending on the type of queue.

#### Tests
Added Unit tests
Ran the regular CI tests as well which would exercise this path through the PlayerController.
Ran around with a character in the TeleportingGym.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/Interop/RPCs/RPC_RingBufferWithACK_Receiver.h [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+template <typename PayloadType, template <typename> class PayloadWrapper, typename SerializerType>
+class MonotonicRingBufferWithACKReceiver : public TRPCBufferReceiver<PayloadType, PayloadWrapper>
+{
+public:
+	MonotonicRingBufferWithACKReceiver(SerializerType&& InSerializer, int32 InNumberOfSlots, PayloadWrapper<PayloadType>&& InWrapper)
+		: TRPCBufferReceiver<PayloadType, PayloadWrapper>(MoveTemp(InWrapper))
+		, Serializer(MoveTemp(InSerializer))
+		, NumberOfSlots(InNumberOfSlots)
+	{
+		ComponentsToRead.Add(Serializer.GetComponentId());
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/RPCs/RPC_RingBufferWithACK_Receiver.h:19]: Unimportant, but is this genericness useful in other places? This class seems to not rely on `RPCBufferReceiver::OnAdded` for reading components, and might be more concise if it did
```
OnAdded(...)
{
	const ComponentData* RpcComponentData = Element.Components.FindByPredicate(ComponentIdEquality{Serializer.GetComponentId()};
	const ComponentData* AckComponentData = Element.Components.FindByPredicate(ComponentIdEquality{Serializer.GetACKComponentId()};
	OnAdded_ReadAckComponent(AckComponentData);
	OnAdded_ReadRpcComponent(RpcComponentData);
}
```

**Review by improbable-dmitrii — COMMENTED** [review]
Good changes, don't have any strong objections or opinions on this PR.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/RPCs/RPC_RingBufferWithACK_Sender.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		int32 AvailableSlots = FMath::Max(0, int32(NumberOfSlots) - int32(NextSlot.CountWritten - NextSlot.LastACK));
+
+		int32 RPCsToWrite = FMath::Min(RPCs.Num(), AvailableSlots);
+		if (RPCsToWrite > 0)
+		{
+			auto EntityWrite = Ctx.WriteTo(EntityId, Serializer.GetComponentId());
+			for (int32 i = 0; i < RPCsToWrite; ++i)
+			{
+				uint64 RPCId = ++NextSlot.CountWritten;
+				uint64 Slot = (RPCId - 1) % NumberOfSlots;
```
[the remark this answers is not in the document]
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/RPCs/RPC_RingBufferWithACK_Sender.h:78]: What do you mean by the "first" ring buffer field ?

**Review by improbable-dmitrii — APPROVED** [review]
Nice!

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCs/RPCRingBufferTest.cpp [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+#include "Tests/TestDefinitions.h"
+
+#include "Interop/RPCs/RPCQueues.h"
+#include "Interop/RPCs/RPC_RingBufferWithACK_Receiver.h"
+#include "Interop/RPCs/RPC_RingBufferWithACK_Sender.h"
+
+#define RPCRINGBUFFER_TEST(TestName) GDK_TEST(Core, RPCRingBuffer, TestName)
+
+using namespace SpatialGDK;
```
[the remark this answers is not in the document]
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/RPCs/RPCRingBufferTest.cpp:14]: That would actually require a private namespace.

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCs/RPCRingBufferTest.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+	EntityComponentMock Data;
+
+	Sender ServerWorker;
+	Receiver ClientWorker;
+	UnboundedQueue ServerQueue;
+
+	RPCRingBufferTest_Fixture(uint32 BufferSize)
+		: ServerWorker(Serializer(Data, Serializer::Writer), BufferSize)
+		, ClientWorker(Serializer(Data, Serializer::Reader), BufferSize, NullReceiveWrapper<Payload>())
+		, ServerQueue(FName(TEXT("DummyQueue")), ServerWorker)
```
[the remark this answers is not in the document]
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/RPCs/RPCRingBufferTest.cpp:120]: I prefer explicit over implicit in that case.

**Pull request merged**